wellcomeml/ml/clustering.py

```diff
@@ -0,0 +1,321 @@
+
+                    n_clusters -= 1*(noise > 0)
+                    params = {
+                        'vectorizer': vectorizer_param,
+                        'reducer': reducer_param,
+                        'clustering': clustering_param
+                    }
+                    params_list += [{
+                        "params": params,
+                        "silhouette": silhouette,
+                        "n_clusters": n_clusters,
+                        "noise": noise
+                    }]
+
+                    if silhouette > best_silhouette and min_n_clusters <= \
+                            n_clusters <= max_n_clusters and noise < max_noise:
+                        best_silhouette = silhouette
+                        best_params = deepcopy(params)
+
+        # Fits the pipeline again with the best parameters
+        self.set_params(best_params)
+        self.fit(X)
+
+        logger.setLevel(logging_level)
+
+        return {
+            "params_list": params_list,
+            "silhouette": best_silhouette,
+            "best_params": best_params
+        }
+
+    def stability(self):
+        """Function to calculate how stable the clusters are"""
+        raise NotImplementedError
+
+    def break_cluster(self, cluster_id):
+        """Breaks a specific by re-clustering a subset"""
+        raise NotImplementedError
+
+    def merge_clusters(self, dictionary):
+        """Merge cluster based on ids"""
+        self.cluster_ids = [dictionary.get(cluster_id, cluster_id) for
+                            cluster_id in self.cluster_ids]
+
+    def set_params(self, params, from_parameter_grid=False):
+        """
+        Set parameters of clustering, reducer and vectorizer
+
+        Args:
+            params: A dictionary of parameters
+            from_parameter_grid: Whether the dictionary is in the form of
+            sklearn parameter grid (e.g. {'clustering__eps': 1}), or a two
+            level dictionary (e.g. {'clustering': {'eps': 1}}) (default).
+
+        Returns:
+            None
+
+        """
+        self.vectorizer.set_params(**params.get('vectorizer', {}))
+        self.reducer.set_params(**params.get('reducer', {}))
+        self.clustering.set_params(**params.get('clustering', {}))
+
+    def _find_keywords(self, X_transformed, n_kw=10):
+        if self.embedding != 'tf-idf':
+            raise ValueError(f'Cannot find keyword for '
+                             f'{self.embedding} vectorizer')
+        idx_to_word = {value: key for key, value in
+                       self.vectorizer.vectorizer.vocabulary_.items()}
+
+        # Initialise a dictionary to calculate the aggregated vector for each
+        # cluster
+
+        aggregated = {key: np.zeros(X_transformed.shape[1])
+                      for key in self.cluster_ids}
+
+        for i in range(len(X_transformed)):
+            aggregated[self.cluster_ids[i]] += X_transformed[i]
```

Comment:
this is a hack right? you are adding the tfidf values for all the examples of a cluster. actually the addition might make some sense since the denominator for each term i.e. the idf score of the term is the same so it would be a combined count divided by idf for the term approx. one things to note here is that this is a particular problem of dbscan though as kmeans, nmf and lda will come up with a prototype representation for each cluster which translates to some words easily. so maybe it would be good to add an if then else to check for dbscan and add not implemented yet for kmeans, nmf, lda etc